src/Private/Dump.ps1

```diff
@@ -0,0 +1,606 @@
+
+.NOTES
+    This function is useful for logging, debugging, and safely displaying complex objects in a readable format.
+
+#>
+function ConvertTo-PodeSerializableObject {
+    param (
+        [object]
+        $InputObject,
+
+        [int]
+        $MaxDepth = 5,
+
+        [int]
+        $CurrentDepth = 0
+    )
+
+    # Check if the current depth has reached or exceeded the maximum allowed depth
+    if ($CurrentDepth -ge $MaxDepth) {
+        # Return a simple message indicating that the maximum depth has been reached
+        return 'Max depth reached'
+    }
+
+    # Handle null input
+    if ($null -eq $InputObject) {
+        return $null  # Return null if the input object is null
+    }
+    # Handle hashtables
+    elseif ($InputObject -is [hashtable]) {
+        $result = @{}
+        try {
+            foreach ($key in $InputObject.Keys) {
+                try {
+                    # Serialize each key-value pair in the hashtable
+                    $strKey = $key.ToString()
+                    $result[$strKey] = ConvertTo-PodeSerializableObject -InputObject $InputObject[$key] -MaxDepth $MaxDepth -CurrentDepth ($CurrentDepth + 1)
+                }
+                catch {
+                    Write-PodeHost $_ -ForegroundColor Red
+                }
+            }
+        }
+        catch {
+            Write-PodeHost $_ -ForegroundColor Red
+        }
+        return $result
+    }
+    # Handle PSCustomObjects
+    elseif ($InputObject -is [PSCustomObject]) {
+        $result = @{}
+        try {
+            foreach ($property in $InputObject.PSObject.Properties) {
+                try {
+                    # Serialize each property in the PSCustomObject
+                    $result[$property.Name.ToString()] = ConvertTo-PodeSerializableObject -InputObject $property.Value -MaxDepth $MaxDepth -CurrentDepth ($CurrentDepth + 1)
+                }
+                catch {
+                    Write-PodeHost $_ -ForegroundColor Red
+                }
+            }
+        }
+        catch {
+            Write-PodeHost $_ -ForegroundColor Red
+        }
+        return $result
+    }
+    # Handle enumerable collections, excluding strings
+    elseif ($InputObject -is [System.Collections.IEnumerable] -and -not ($InputObject -is [string])) {
+        # Serialize each item in the collection
+        return $InputObject | ForEach-Object { ConvertTo-PodeSerializableObject -InputObject $_ -MaxDepth $MaxDepth -CurrentDepth ($CurrentDepth + 1) }
+    }
+    else {
+        # Convert other object types to string for serialization
+        return $InputObject.ToString()
+    }
+}
+
+function Initialize-PodeDebugHandler {
+    # Embed C# code to handle the DebuggerStop event
+    Add-Type @'
+using System;
+using System.Management.Automation;
+using System.Management.Automation.Runspaces;
+using System.Collections.ObjectModel;
+
+namespace Pode.Embedded
+{
+    public class DebuggerHandler
```

Comment:
I started testing and hit the first issue with this class, since all of the properties and methods are static it causes subsequent calls after the first to become contaminated - for example the first runspace sets `isEventTriggered` to `true`, but then it's true for all other runspaces even though it's actually false.

I fixed the C# code to be the following further below - including updating how the Dump works to below that. BUT, that's when I hit the first major snag, with the class fixed and using `Web-Dump.ps1` example, all of the "waiting for debugs" timeout after 60s - and they will for basically ever until several appropriate requests hit the server as Pode is waiting for inbound requests.

![image](https://github.com/user-attachments/assets/219d5fbd-1322-4144-a6da-74d1e8b20b26)

Specifically it's lines like below which hang waiting, in `PodeServer.ps1`:
```powershell
$context = (Wait-PodeTask -Task $Listener.GetContextAsync($PodeContext.Tokens.Cancellation.Token))
```

I'm trying to think of ways around this, and one idea I had was to give each Runspace its own CancellationToken and this is used on `.GetContextAsync` calls. When we try to attach the debugger to a runspace, we `Enable-RunspaceDebug` and then Cancel that runspace's CancellationToken, and then renew it for the next loop run - that "should" stop the timeouts.

When the server stops/restarts, then we need to Cancel all the runspace tokens as well.

### Fix DebuggerHandler
```c#
using System;
using System.Management.Automation;
using System.Management.Automation.Runspaces;
using System.Collections.ObjectModel;

namespace Pode.Embedded
{
    public class DebuggerHandler : IDisposable
    {
        // Collection to store variables collected during the debugging session
        public PSDataCollection<PSObject> Variables { get; private set; } = new PSDataCollection<PSObject>();

        // Event handler for the DebuggerStop event
        private EventHandler<DebuggerStopEventArgs> DebuggerStopHandler;

        // Flag to indicate whether the DebuggerStop event has been triggered
        public bool IsEventTriggered { get; private set; } = false;

        // Flag to control whether variables should be collected during the DebuggerStop event
        private bool CollectVariables = true;

        // Runspace object to store the runspace that the debugger is attached to
        private Runspace Runspace;

        public DebuggerHandler(Runspace runspace, bool collectVariables = true)
        {
            // Set the collection flag and Runspace object
            CollectVariables = collectVariables;
            Runspace = runspace;

            // Initialize the event handler with the OnDebuggerStop method
            DebuggerStopHandler = new EventHandler<DebuggerStopEventArgs>(OnDebuggerStop);
            Runspace.Debugger.DebuggerStop += DebuggerStopHandler;
        }

        // Method to detach the DebuggerStop event handler from the runspace's debugger, and general clean-up
        public void Dispose()
        {
            IsEventTriggered = false;

            // Remove the event handler to prevent further event handling
            if (DebuggerStopHandler != default(EventHandler<DebuggerStopEventArgs>))
            {
                Runspace.Debugger.DebuggerStop -= DebuggerStopHandler;
                DebuggerStopHandler = null;
            }

            // Clean-up variables
            Runspace = default(Runspace);
            Variables.Clear();

            // Garbage collection
            GC.SuppressFinalize(this);
        }

        // Event handler method that gets called when the debugger stops
        private void OnDebuggerStop(object sender, DebuggerStopEventArgs args)
        {
            // Set the eventTriggered flag to true
            IsEventTriggered = true;

            // Cast the sender to a Debugger object
            var debugger = sender as Debugger;
            if (debugger == default(Debugger))
            {
                return;
            }

            // Enable step mode to allow for command execution during the debug stop
            debugger.SetDebuggerStepMode(true);

            // Collect variables or hang the debugger
            var command = new PSCommand();
            command.AddCommand(CollectVariables
                ? "Get-PodeDumpScopedVariable"
                : "while($PodeContext.Server.Suspended) { Start-Sleep -Milliseconds 500 }");

            // Execute the command within the debugger
            var outputCollection = new PSDataCollection<PSObject>();
            debugger.ProcessCommand(command, outputCollection);

            // Add results to the variables collection if collecting variables
            if (CollectVariables)
            {
                foreach (var output in outputCollection)
                {
                    Variables.Add(output);
                }
            }
            else
            {
                // Ensure the debugger remains ready for further interaction
                debugger.SetDebuggerStepMode(true);
            }
        }
    }
}
```

### Updated Get-PodeRunspaceVariablesViaDebugger
```powershell
function Get-PodeRunspaceVariablesViaDebugger {
    param (
        [Parameter(Mandatory)]
        [System.Management.Automation.Runspaces.Runspace]$Runspace,

        [Parameter()]
        [int]$Timeout = 60
    )

    # Initialize variables collection
    $variables = @()
    try {

        # Attach the debugger and break all
        $debugger = [Pode.Embedded.DebuggerHandler]::new($Runspace)
        Enable-RunspaceDebug -BreakAll -Runspace $Runspace

        # Wait for the event to be triggered or timeout
        $startTime = [DateTime]::UtcNow
        Write-PodeHost "Waiting for $($Runspace.Name) to enter in debug ." -NoNewLine

        while (!$debugger.IsEventTriggered) {
            Start-Sleep -Milliseconds 500
            Write-PodeHost '.' -NoNewLine

            if (([DateTime]::UtcNow - $startTime).TotalSeconds -ge $Timeout) {
                Write-PodeHost "Failed (Timeout reached after $Timeout seconds.)"
                return @{}
            }
        }

        Write-PodeHost 'Done'

        # Retrieve and output the collected variables from the embedded C# code
        $variables = $debugger.Variables
    }
    catch {
        # Log the error details using Write-PodeErrorLog.
        # This ensures that any exceptions thrown during the execution are logged appropriately.
        $_ | Write-PodeErrorLog
    }
    finally {
        # Detach the debugger from the runspace to clean up resources and prevent any lingering event handlers.
        if ($null -ne $debugger) {
            $debugger.Dispose()
        }

        # Disable debugging for the runspace. This ensures that the runspace returns to its normal execution state.
        Disable-RunspaceDebug -Runspace $Runspace
    }

    return $variables[0]
}
```

Reply:
I never encountered this issue before. But your implementation is better